src/data/nat/digits.lean

```diff
@@ -0,0 +1,346 @@
+  ring,
+end
+
+lemma dvd_of_digits_sub_of_digits {α : Type*} [euclidean_domain α]
+  {a b k : α} (h : k ∣ a - b) (L : list ℕ) :
+  k ∣ ((of_digits a L) - (of_digits b L)) :=
+begin
+  induction L,
+  { change k ∣ 0 - 0, simp, },
+  { dsimp [of_digits],
+    simp only [add_sub_add_left_eq_sub],
+    exact dvd_mul_sub_mul h L_ih, }
+end
+
+lemma of_digits_modeq' (b b' : ℕ) (k : ℕ) (h : b ≡ b' [MOD k]) (L : list ℕ) : (of_digits b L) ≡ (of_digits b' L) [MOD k] :=
+begin
+  induction L,
+  { refl, },
+  { dsimp [of_digits],
+    dsimp [nat.modeq] at *,
+    conv_lhs { rw [nat.add_mod, nat.mul_mod, h, L_ih], },
+    conv_rhs { rw [nat.add_mod, nat.mul_mod], }, }
+end
+
+lemma of_digits_modeq (b k : ℕ) (L : list ℕ) : (of_digits b L) ≡ (of_digits (b % k) L) [MOD k] :=
+of_digits_modeq' b (b % k) k (nat.modeq.symm (nat.modeq.mod_modeq b k)) L
+
+lemma of_digits_mod (b k : ℕ) (L : list ℕ) : (of_digits b L) % k = (of_digits (b % k) L) % k :=
+of_digits_modeq b k L
+
+lemma of_digits_zmodeq' (b b' : ℤ) (k : ℕ) (h : b ≡ b' [ZMOD k]) (L : list ℕ) : (of_digits b L) ≡ (of_digits b' L) [ZMOD k] :=
+begin
+  induction L,
+  { refl, },
+  { dsimp [of_digits],
+    dsimp [int.modeq] at *,
+    conv_lhs { rw [int.add_mod, int.mul_mod, h, L_ih], },
+    conv_rhs { rw [int.add_mod, int.mul_mod], }, }
+end
+
+lemma of_digits_zmodeq (b : ℤ) (k : ℕ) (L : list ℕ) : (of_digits b L) ≡ (of_digits (b % k) L) [ZMOD k] :=
+of_digits_zmodeq' b (b % k) k (int.modeq.symm (int.modeq.mod_modeq b ↑k)) L
+
+lemma of_digits_zmod (b : ℤ) (k : ℕ) (L : list ℕ) : (of_digits b L) % k = (of_digits (b % k) L) % k :=
+of_digits_zmodeq b k L
+
+lemma modeq_digits_sum (b b' : ℕ) (h : b' % b = 1) (n : ℕ) :
+   n ≡ (digits b' n).sum [MOD b] :=
+begin
+  rw ←of_digits_one,
+  conv { congr, skip, rw ←(of_digits_digits b' n) },
+  convert of_digits_modeq _ _ _,
+  exact h.symm,
+end
+
+lemma modeq_three_digits_sum (n : ℕ) : n ≡ (digits 10 n).sum [MOD 3] :=
+modeq_digits_sum 3 10 (by norm_num) n
+
+lemma modeq_nine_digits_sum (n : ℕ) : n ≡ (digits 10 n).sum [MOD 9] :=
+modeq_digits_sum 9 10 (by norm_num) n
+
+lemma dvd_iff_dvd_digits_sum (b b' : ℕ) (h : b' % b = 1) (n : ℕ) :
+  b ∣ n ↔ b ∣ (digits b' n).sum :=
+begin
+  rw ←of_digits_one,
+  conv_lhs { rw ←(of_digits_digits b' n) },
+  rw [nat.dvd_iff_mod_eq_zero, nat.dvd_iff_mod_eq_zero, of_digits_mod, h],
+end
+
+lemma three_dvd_iff (n : ℕ) : 3 ∣ n ↔ 3 ∣ (digits 10 n).sum :=
+dvd_iff_dvd_digits_sum 3 10 (by norm_num) n
+
+lemma nine_dvd_iff (n : ℕ) : 9 ∣ n ↔ 9 ∣ (digits 10 n).sum :=
+dvd_iff_dvd_digits_sum 9 10 (by norm_num) n
+
+lemma zmodeq_of_digits_digits (b b' : ℕ) (c : ℤ) (h : b' ≡ c [ZMOD b]) (n : ℕ) :
+  n ≡ (of_digits c (digits b' n)) [ZMOD b] :=
+begin
+  conv { congr, skip, rw ←(of_digits_digits b' n) },
+  rw coe_int_of_digits,
+  apply of_digits_zmodeq' _ _ _ h,
+end
+
+lemma modeq_eleven_digits_sum (n : ℕ) : n ≡ of_digits (-1 : ℤ) (digits 10 n) [ZMOD 11] :=
+zmodeq_of_digits_digits 11 10 (-1 : ℤ) dec_trivial n
+
+-- TODO move?
```

Comment:
Yes, please move it (and generalize it to any `comm_ring`).

Reply:
Even `ring`, right?

Reply:
You need to have a divisibility notion, which is not an instance in rings as far as I can tell (you would have to choose between left and right divisibility)

Reply:
Hmm, that's true. Ok, `comm_ring` it will be (-;

Reply:
Done.